src/cljam/io/wig.clj

```diff
@@ -0,0 +1,171 @@
+(ns cljam.io.wig
+  "Functions to read and write the WIG (Wiggle) format. See
+  https://genome.ucsc.edu/goldenpath/help/wiggle.html for the detail WIG
+  specifications."
+  (:require [clojure.java.io :as cio]
+            [clojure.string :as cstr]
+            [proton.core :refer [as-long as-double]]
+            [cljam.io.protocols :as protocols]
+            [cljam.util :as util]
+            [cljam.util.chromosome :as chr])
+  (:import [java.net URL]
+           [java.io BufferedReader BufferedWriter Closeable]))
+
+(declare read-fields write-fields)
+
+(defrecord WIGReader [^BufferedReader reader ^URL url]
+  Closeable
+  (close [this]
+    (.close ^Closeable (.reader this)))
+  protocols/IReader
+  (reader-url [this] (.url this))
+  (read [this] (read-fields this))
+  (read [this option] (read-fields this))
+  (indexed? [_] false))
+
+(defrecord WIGWriter [^BufferedWriter writer ^URL url]
+  Closeable
+  (close [this]
+    (.close ^Closeable (.writer this)))
+  protocols/IWriter
+  (writer-url [this] (.url this)))
+
+(defn ^WIGReader reader
+  "Returns an open cljam.io.wig.WIGReader of f. Should be used inside with-open
+  to ensure the reader is properly closed."
+  [f]
+  (WIGReader. (cio/reader (util/compressor-input-stream f)) (util/as-url f)))
+
+(defn ^WIGWriter writer
+  "Returns an open cljam.io.wig.WIGWriter of f. Should be used inside with-open
+  to ensure the writer is properly closed."
+  [f]
+  (WIGWriter. (cio/writer (util/compressor-output-stream f)) (util/as-url f)))
+
+(defn- header-or-comment?
+  "Checks if given string is neither a header nor a comment line."
+  [^String s]
+  (or (empty? s)
+      (.startsWith s "browser")
+      (.startsWith s "track")
+      (.startsWith s "#")))
+
+(defn- normalize
+  "Normalize WIG lines."
+  [m]
+  (update m :chrom chr/normalize-chromosome-key))
+
+(defn- fields->map
+  "Convert vector [\"key1=value1\" \"key2=value2\" ...] to map {:key1 \"value1\" :key2 \"value2\" ...}."
+  [fields]
+  (->> fields
+       (map #(cstr/split % #"="))
+       (map (fn [[k v]] [(keyword k) v]))
+       (into {})))
+
+(defn- str->wiggle-track-data
+  "Convert string to Wiggle track data values which can be integer or real,
+  positive or negative values."
+  [^String s]
+  (if-let [l (as-long s)]
+    l
+    (as-double s)))
+
+(defn- deserialize-wigs
+  "Parse WIG lines and returns a vector of maps."
+  [lines]
+  (loop [lines lines
```

Comment:
Nested lazy sequences, in general, are hard to use correctly and I don't think it's a good idea to choose the second option.

As for Option 1. and 3., I think both readers would be useful in different use cases, so I would like both if possible 😄

Reply:
Thank you for the detailed comments. I'll choose Option 3. for memory.

Then, I'd like to add a `:format` key to flat map, because a writer cannot determine format without the key. For example, if a writer tries to output this data,
```
({:chr "chr1", :start 1, :end 10, :value 1}
 {:chr "chr1", :start 100, :end 110, :value 2})
```
cannot determine which formats to output.

```
variableStep chrom=chr1 span=10
1 1
100 2
```
```
fixedStep chrom=chr1 start=1 step=100 span=10
1
2
```